File: python/python_demo/data_structures/queues/circular_queue_linkedlist_based.py
class Node:
    """A node in the circular linked list."""
    def __init__(self, data):
        self.data = data
        self.next = None

class CircularQueue:
    """Circular queue implementation using a linked list."""
    def __init__(self):
        self.front = self.rear = None

    def enqueue(self, value):
        """Adds an element to the queue."""
        new_node = Node(value)
        if not self.front:
            self.front = self.rear = new_node
            self.rear.next = self.front  # Circular link
        else:
            self.rear.next = new_node
            self.rear = new_node
            self.rear.next = self.front  # Maintain circular connection

    def dequeue(self):
        """Removes and returns the front element from the queue."""
        if not self.front:
            print("Queue is empty!")
            return None
        removed_value = self.front.data
        if self.front == self.rear:  # Single element case
            self.front = self.rear = None
        else:
            self.front = self.front.next
            self.rear.next = self.front  # Maintain circular connection
        return removed_value

    def display(self):
        """Displays the elements of the queue."""
        if not self.front:
            print("Queue is empty!")
            return
        temp = self.front
        while True:
            print(temp.data, end=" ")
            temp = temp.next
            if temp == self.front:
                break
        print()

def demo_circular_queue_linkedlist_based():
    print("==================================")
    print("Demo of linkedlist based circular queue")
    print("==================================")
    # Demo
    queue = CircularQueue()
    queue.enqueue(10)
    queue.enqueue(20)
    queue.enqueue(30)
    queue.display()  # Output: 10 20 30
    print("Dequeued:", queue.dequeue())  # Output: 10
    queue.display()  # Output: 20 30

File: java/java_demo/src/data_structures/queues/CircularQueueLinkedListBased.java
package data_structures.queues;

public class CircularQueueLinkedListBased {
    private static class Node {
        int data;
        Node next;
        public Node(int data) { this.data = data; }
    }

    private Node front, rear;

    /**
     * Inserts an element into the circular queue.
     * @param value Value to be added.
     */
    public void enqueue(int value) {
        Node newNode = new Node(value);
        if (front == null) {
            front = rear = newNode;
            rear.next = front;  // Circular link
        } else {
            rear.next = newNode;
            rear = newNode;
            rear.next = front;  // Maintain circular connection
        }
    }

    /**
     * Removes and returns the front element from the queue.
     * @return The removed element, or -1 if the queue is empty.
     */
    public int dequeue() {
        if (front == null) {
            System.out.println("Queue is empty!");
            return -1;
        }
        int removedValue = front.data;
        if (front == rear) { // Single element case
            front = rear = null;
        } else {
            front = front.next;
            rear.next = front;  // Maintain circular connection
        }
        return removedValue;
    }

    /**
     * Displays the elements of the queue.
     */
    public void display() {
        if (front == null) {
            System.out.println("Queue is empty!");
            return;
        }
        Node temp = front;
        do {
            System.out.print(temp.data + " ");
            temp = temp.next;
        } while (temp != front);
        System.out.println();
    }

    public static void demo() {
        System.out.println("===========================");
        System.out.println("Circular Queue demo based on LinkedList");
        System.out.println("===========================");
        CircularQueueLinkedListBased queue = new CircularQueueLinkedListBased();
        queue.enqueue(10);
        queue.enqueue(20);
        queue.enqueue(30);
        queue.display();  // Output: 10 20 30
        System.out.println("Dequeued: " + queue.dequeue());  // Output: 10
        queue.display();  // Output: 20 30
    }
}
